File: src/test/java/org/dicio/skill/chain/ChainSkillTest.java
package org.dicio.skill.chain;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertSame;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import org.dicio.skill.Skill;
import org.dicio.skill.SkillComponent;
import org.dicio.skill.SkillContext;
import org.dicio.skill.SkillInfo;
import org.junit.Test;

import java.util.List;

public class ChainSkillTest {

    private static final SkillInfo skillInfo = new SkillInfo("", 0, 0, 0, false) {
        @Override public boolean isAvailable(final SkillContext context) { return false; }
        @Override public Skill build(final SkillContext context) { return null; }
        @Nullable @Override public Fragment getPreferenceFragment() { return null; }
    };

    // the type parameters used here are just random: the chain skill does not check if they match

    private static final InputRecognizer<Integer> ir = new InputRecognizer<Integer>() {
        @Override public Specificity specificity() { return Specificity.HIGH; }
        @Override public void setInput(final String input, final List<String> inputWords,
                final List<String> normalizedInputWords) {}
        @Override public float score() { return 0; }
        @Override public Integer getResult() { return 8; }
        @Override public void cleanup() {}
    };

    private static final IntermediateProcessor<Integer, Double> ip1
            = new IntermediateProcessor<Integer, Double>() {
        @Override public Double process(final Integer data) { return data / 3.0; }
    };

    private static final IntermediateProcessor<Double, Float> ip2
            = new IntermediateProcessor<Double, Float>() {
        @Override public Float process(final Double data) { return (float)(2 * data); }
    };

    private static final StringBuilder generatedOutput = new StringBuilder();
    private static final OutputGenerator<Object> og
            = new OutputGenerator<Object>() {
        @Override
        public void generate(final Object data) {
            assertNotNull(data);
            generatedOutput.append(data.getClass()).append("-").append(data);
        }
        @Override public void cleanup() {}
    };

    private static void assertGeneratedOutput(final String expected, final Skill skill)
            throws Exception {
        // use static generatedOutput variable
        generatedOutput.setLength(0);
        skill.processInput();
        skill.generateOutput();
        assertEquals(expected, generatedOutput.toString());
    }

    private static void assertSetSkillInfo(final Skill skill, final SkillComponent... skillComponents) {
        skill.setSkillInfo(null);
        for (final SkillComponent skillComponent : skillComponents) {
            skillComponent.setSkillInfo(null);
        }

        skill.setSkillInfo(skillInfo);

        assertSame(skillInfo, skill.getSkillInfo());
        for (final SkillComponent skillComponent : skillComponents) {
            assertSame(skillInfo, skillComponent.getSkillInfo());
        }
    }


    @Test
    public void testBuildNoIntermediateProcessors() throws Exception {
        final Skill skill = new ChainSkill.Builder(ir)
                .output(og);

        assertSetSkillInfo(skill, ir, og);
        assertEquals(ir.specificity(), skill.specificity());
        assertGeneratedOutput(Integer.class + "-" + 8, skill);
    }

    @Test
    public void testBuildOneIntermediateProcessor() throws Exception {
        final Skill skill = new ChainSkill.Builder(ir)
                .process(ip1)
                .output(og);

        assertSetSkillInfo(skill, ir, ip1, og);
        assertEquals(ir.specificity(), skill.specificity());
        assertGeneratedOutput(Double.class + "-" + (8 / 3.0), skill);
    }

    @Test
    public void testBuildTwoIntermediateProcessors() throws Exception {
        final Skill skill = new ChainSkill.Builder(ir)
                .process(ip1)
                .process(ip2)
                .output(og);

        assertSetSkillInfo(skill, ir, ip1, ip2, og);
        assertEquals(ir.specificity(), skill.specificity());
        assertGeneratedOutput(Float.class + "-" + (float)(2 * (8 / 3.0)), skill);
    }
}

File: src/test/java/org/dicio/skill/chain/ChainSkillTest.kt
package org.dicio.skill.chain

import androidx.compose.runtime.Composable
import androidx.fragment.app.Fragment
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeSameInstanceAs
import org.dicio.skill.Skill
import org.dicio.skill.SkillComponent
import org.dicio.skill.SkillContext
import org.dicio.skill.SkillInfo
import org.dicio.skill.output.SkillOutput

class ChainSkillTest : StringSpec({
    "build, no intermediate processor" {
        val skill: Skill = ChainSkill.Builder(TestSkillInfo, ir)
            .output(og)

        skill.specificity shouldBeSameInstanceAs ir.specificity
        assertGeneratedOutput(java.lang.Integer::class.java.toString() + "|" + 8, skill)
    }

    "build, one intermediate processor" {
        val skill: Skill = ChainSkill.Builder(TestSkillInfo, ir)
            .process(ip1)
            .output(og)

        skill.specificity shouldBeSameInstanceAs ir.specificity
        assertGeneratedOutput(java.lang.Double::class.java.toString() + "|" + (8 / 3.0), skill)
    }

    "build, two intermediate processors" {
        val skill: Skill = ChainSkill.Builder(TestSkillInfo, ir)
            .process(ip1)
            .process(ip2)
            .output(og)

        skill.specificity shouldBeSameInstanceAs ir.specificity
        assertGeneratedOutput(java.lang.Float::class.java.toString() + "|" + (2 * (8 / 3.0)).toFloat(), skill)
    }
})

object TestSkillInfo : SkillInfo("", 0, 0, 0, false) {
    override fun isAvailable(context: SkillContext) = TODO()
    override fun build(context: SkillContext) = TODO()
    override val preferenceFragment get() = TODO()
}

private val skillInfo: SkillInfo = object : SkillInfo("", 0, 0, 0, false) {
    override fun isAvailable(context: SkillContext): Boolean {
        return false
    }

    override fun build(context: SkillContext): Skill {
        throw NotImplementedError()
    }

    override val preferenceFragment: Fragment?
        get() = null
}

// the type parameters used here are just random: the chain skill does not check if they match
private val ir = object : InputRecognizer<Int>(Specificity.HIGH) {
    override fun setInput(
        input: String, inputWords: List<String>,
        normalizedInputWords: List<String>
    ) {
    }

    override fun score(): Float {
        return 0.0f
    }

    override val result: Int
        get() = 8

    override fun cleanup() {}
}

private val ip1 = object : IntermediateProcessor<Int, Double>() {
    override fun process(data: Int): Double {
        return data / 3.0
    }
}

private val ip2 = object : IntermediateProcessor<Double, Float>() {
    override fun process(data: Double): Float {
        return (2 * data).toFloat()
    }
}

private val og = object : OutputGenerator<Any>() {
    override fun generate(data: Any): SkillOutput {
        data.shouldNotBeNull()
        return object : SkillOutput {
            override fun getSpeechOutput(ctx: SkillContext): String = "${data.javaClass}|$data"
            @Composable override fun GraphicalOutput(ctx: SkillContext) {}
        }
    }

    override fun cleanup() {}
}

@Throws(Exception::class)
private fun assertGeneratedOutput(expected: String, skill: Skill) {
    skill.processInput()
    skill.generateOutput().getSpeechOutput(SkillContext()) shouldBe expected
}
